File: app/src/main/java/com/mvcoding/financius/ui/calculator/CalculatorPresenter.java
/*
 * Copyright (C) 2015 Mantas Varnagiris.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 */

package com.mvcoding.financius.ui.calculator;

import android.support.annotation.NonNull;

import com.mvcoding.financius.ui.ActivityScope;
import com.mvcoding.financius.ui.Presenter;
import com.mvcoding.financius.ui.PresenterView;

import java.math.BigDecimal;
import java.util.ArrayList;
import java.util.List;

import javax.inject.Inject;

import rx.Observable;

@ActivityScope class CalculatorPresenter extends Presenter<CalculatorPresenter.View> {
    private final Calculator calculator;

    @Inject CalculatorPresenter(@NonNull Calculator calculator) {
        this.calculator = calculator;
    }

    @Override protected void onViewAttached(@NonNull View view) {
        super.onViewAttached(view);

        final Observable<Object> equalsObservable = view.onEquals();
        final Observable<Object> numberObservable = view.on0Number()
                .doOnNext(o -> calculator.digit0())
                .mergeWith(view.on1Number().doOnNext(o -> calculator.digit1()))
                .mergeWith(view.on2Number().doOnNext(o -> calculator.digit2()))
                .mergeWith(view.on3Number().doOnNext(o -> calculator.digit3()))
                .mergeWith(view.on4Number().doOnNext(o -> calculator.digit4()))
                .mergeWith(view.on5Number().doOnNext(o -> calculator.digit5()))
                .mergeWith(view.on6Number().doOnNext(o -> calculator.digit6()))
                .mergeWith(view.on7Number().doOnNext(o -> calculator.digit7()))
                .mergeWith(view.on8Number().doOnNext(o -> calculator.digit8()))
                .mergeWith(view.on9Number().doOnNext(o -> calculator.digit9()));

        final List<Observable<Object>> expressionUpdates = new ArrayList<>();
        expressionUpdates.add(numberObservable);
        expressionUpdates.add(view.onDecimal().doOnNext(o -> calculator.decimal()));
        expressionUpdates.add(view.onAdd().doOnNext(o -> calculator.add()));
        expressionUpdates.add(view.onSubtract().doOnNext(o -> calculator.subtract()));
        expressionUpdates.add(view.onMultiply().doOnNext(o -> calculator.multiply()));
        expressionUpdates.add(view.onDivide().doOnNext(o -> calculator.divide()));
        expressionUpdates.add(view.onDelete().doOnNext(o -> calculator.delete()));
        expressionUpdates.add(view.onClear().doOnNext(o -> calculator.clear()).doOnNext(o -> view.clearExpression()));
        expressionUpdates.add(view.onNumberChange().doOnNext(calculator::setNumber).map(number -> new Object()));
        expressionUpdates.add(equalsObservable.filter(o -> !calculator.isEmptyOrSingleNumber())
                                      .map(o -> calculator.calculate())
                                      .doOnNext(calculator::setNumber)
                                      .map(number -> new Object()));
        expressionUpdates.add(calculator.getExpression().isEmpty() ? Observable.empty() : Observable.just(new Object()));

        unsubscribeOnDetach(Observable.merge(expressionUpdates)
                                    .map(o -> calculator.getExpression())
                                    .doOnNext(view::showExpression)
                                    .subscribe(expression -> {
                                        if (calculator.isEmptyOrSingleNumber()) {
                                            view.showStartResult();
                                        } else {
                                            view.showCalculate();
                                        }
                                    }));

        unsubscribeOnDetach(equalsObservable.filter(o -> calculator.isEmptyOrSingleNumber())
                                    .map(o -> calculator.calculate())
                                    .subscribe(view::startResult));
    }

    public interface View extends PresenterView {
        @NonNull Observable<Object> on0Number();
        @NonNull Observable<Object> on1Number();
        @NonNull Observable<Object> on2Number();
        @NonNull Observable<Object> on3Number();
        @NonNull Observable<Object> on4Number();
        @NonNull Observable<Object> on5Number();
        @NonNull Observable<Object> on6Number();
        @NonNull Observable<Object> on7Number();
        @NonNull Observable<Object> on8Number();
        @NonNull Observable<Object> on9Number();
        @NonNull Observable<Object> onDecimal();
        @NonNull Observable<Object> onEquals();
        @NonNull Observable<Object> onDivide();
        @NonNull Observable<Object> onMultiply();
        @NonNull Observable<Object> onSubtract();
        @NonNull Observable<Object> onAdd();
        @NonNull Observable<Object> onDelete();
        @NonNull Observable<Object> onClear();
        @NonNull Observable<BigDecimal> onNumberChange();
        void showExpression(@NonNull String expression);
        void clearExpression();
        void showCalculate();
        void showStartResult();
        void startResult(@NonNull BigDecimal result);
    }
}

File: core-app/src/main/kotlin/com/mvcoding/financius/feature/calculator/CalculatorPresenter.kt
/*
 * Copyright (C) 2015 Mantas Varnagiris.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 */

package com.mvcoding.financius.feature.calculator

import com.mvcoding.financius.feature.Presenter
import com.mvcoding.financius.feature.calculator.CalculatorPresenter.State.CALCULATE
import com.mvcoding.financius.feature.calculator.CalculatorPresenter.State.SAVE
import rx.Observable
import rx.Observable.merge
import java.math.BigDecimal

class CalculatorPresenter(
        private val calculator: Calculator,
        private val initialNumber: BigDecimal? = null) : Presenter<CalculatorPresenter.View>() {

    init {
        if (initialNumber != null) {
            calculator.setNumber(initialNumber)
        }
    }

    override fun onAttachView(view: View) {
        super.onAttachView(view)

        val expressionAlteringObservable = merge(arrayOf(
                view.onDigit0().doOnNext { calculator.digit0() },
                view.onDigit1().doOnNext { calculator.digit1() },
                view.onDigit2().doOnNext { calculator.digit2() },
                view.onDigit3().doOnNext { calculator.digit3() },
                view.onDigit4().doOnNext { calculator.digit4() },
                view.onDigit5().doOnNext { calculator.digit5() },
                view.onDigit6().doOnNext { calculator.digit6() },
                view.onDigit7().doOnNext { calculator.digit7() },
                view.onDigit8().doOnNext { calculator.digit8() },
                view.onDigit9().doOnNext { calculator.digit9() },
                view.onDecimal().doOnNext { calculator.decimal() },
                view.onAdd().doOnNext { calculator.add() },
                view.onSubtract().doOnNext { calculator.subtract() },
                view.onMultiply().doOnNext { calculator.multiply() },
                view.onDivide().doOnNext { calculator.divide() },
                view.onClear().doOnNext { calculator.clear() },
                view.onDelete().doOnNext { calculator.delete() },
                view.onCalculate().map { calculator.calculate() }.doOnNext { calculator.setNumber(it) }))
                .startWith(Unit)
                .map { calculator.getExpression() }
                .doOnNext { view.showState(if (calculator.isEmptyOrSingleNumber()) SAVE else CALCULATE) }
                .doOnNext { view.showExpression(it) }
                .map { calculator.calculate() }

        unsubscribeOnDetach(view.onSave()
                .withLatestFrom(expressionAlteringObservable, { unit, number -> number })
                .subscribe { view.startResult(it) })
    }

    enum class State { SAVE, CALCULATE }

    interface View : Presenter.View {
        fun onDigit0(): Observable<Unit>
        fun onDigit1(): Observable<Unit>
        fun onDigit2(): Observable<Unit>
        fun onDigit3(): Observable<Unit>
        fun onDigit4(): Observable<Unit>
        fun onDigit5(): Observable<Unit>
        fun onDigit6(): Observable<Unit>
        fun onDigit7(): Observable<Unit>
        fun onDigit8(): Observable<Unit>
        fun onDigit9(): Observable<Unit>
        fun onDecimal(): Observable<Unit>
        fun onAdd(): Observable<Unit>
        fun onSubtract(): Observable<Unit>
        fun onMultiply(): Observable<Unit>
        fun onDivide(): Observable<Unit>
        fun onDelete(): Observable<Unit>
        fun onClear(): Observable<Unit>
        fun onCalculate(): Observable<Unit>
        fun onSave(): Observable<Unit>
        fun showExpression(expression: String)
        fun showState(state: State)
        fun startResult(number: BigDecimal)
    }
}